File: vn2/make_slicefile.c
#include<stdio.h>
#include<stdlib.h>
#include<string.h>
FILE *filebuf;
void ArgumentError();
void MakeTopology(int width,int height,int slice_unit);
void Add_slice(int argc,char *argv[]);
void Del_slice(int argc,char *argv[]);
void Add_host(int argc,char *argv[]);
void Del_host(int argc,char *argv[]);
void convert_Mac(int Mac[],int hostnum);
FILE* search_slice_id(int slice_id,FILE* point);

int main(int aint rgc, char *argv[]){
  if(argc < 2){
    ArgumentError();
    exit(1);
  }

  if(strcmp(argv[1],"make") == 0){
    if(argc != 5){
      ArgumentError();
      exit(1);
    }
    int height = atoi(argv[2]);
    int width = atoi(argv[3]);
    int slice_unit = atoi(argv[4]);
    MakeTopology(width,height,slice_unit);
  }else{
    if((filebuf = fopen("test_slice.conf","r+")) == NULL){
      printf("error!\n");
      exit(0);
    }
    if(strcmp(argv[1],"add_slice") == 0){
      Add_slice(argc,argv);
    }else if(strcmp(argv[1],"del_slice") == 0){
      Del_slice(argc,argv);
    }else if(strcmp(argv[1],"add_host") == 0){
      Add_host(argc,argv);
    }else if(strcmp(argv[1],"del_host") == 0){
      Del_slice(argc,argv);
    }else{
      ArgumentError();
      exit(1);
    } 
  }
  fclose(filebuf);
  exit(0);
}

FILE* search_slice_id(int slice_id,FILE *point){
  char line[1024];
  char* tok;
  while(fgets(line,1024,point) != NULL){
    tok = strtok(&line[5],"{");
    if(slice_id == atoi(tok)){
      return point;
    }
  }
  return NULL;
}

void convert_Mac(int Mac[],int hostnum){
  int i = 0;
  Mac[0] = hostnum;
  while(Mac[i] > 16){
    while(Mac[i] > 16){
      Mac[i+1]++;
      Mac[i] -= 16;
    }
    i++;
  }
  return;
}

void ArgumentError(){
  printf("usage: ./make_topology make height_nodenumber width_nodenumber slice_unitnum\n");
  printf("usage: ./make_topology add_slice slice_id\n");
  printf("usage: ./make_topology del_slice slice_id\n");
  printf("usage: ./make_topology add_host slice_id host_name1..\n");
  printf("usage: ./make_topology del_host slice_id host_name1..\n");
  return;
}

void MakeTopology(int width,int height,int slice_unit){
  if((filebuf = fopen("test_slice.conf","w")) == NULL){
    printf("error!\n");
    exit(0);
  }

  int size = height*width;
  int i,j,current = 0;
  int mac[12];

  for(i = 0; i < 12; i++){
    mac[i] = 0;
  }

  for(i=0; i<size; i++){
    int slicenum = i / slice_unit;
    if(i%slice_unit == 0){
      fprintf(filebuf, "slice%d{", slicenum);
    }
    for(j=11; j>=0; j--){
      fprintf(filebuf, "%x", mac[j]);
      if((j%2==0) && (j != 0)){
	fprintf(filebuf, ":");
      }
    }
    if(i%slice_unit == slice_unit-1 || i == size-1){
      fprintf(filebuf, "}\n");
    }
    else{
      fprintf(filebuf, ",");
    }

    mac[0]++;
    while(mac[current] == 16){
      mac[current] = 0;
      current++;
      mac[current]++;
    }
    current = 0;
  }
  return;
}

void Add_slice(int argc,char *argv[]){
  char buf[1000000];
  filebuf = search_slice_id(3,filebuf);
  while((getc(filebuf)) != '{');
  
  return;
}

void Del_slice(int argc,char *argv[]){
  return;
}
void Add_host(int argc,char *argv[]){
  return;
}
void Del_host(int argc,char *argv[]){
  return;
}

File: vn2/make_slicefile.pl
#!/usr/bin/perl
use strict;
use warnings;
use utf8;

if(@ARGV < 2){
    print "error\n";
}
if($ARGV[0] eq "make"){
    &make_topology;
}elsif($ARGV[0] eq "add_slice"){
    &add_slice;
}elsif($ARGV[0] eq "add_host"){
    &add_host;
}elsif($ARGV[0] eq "del_slice"){
    &del_slice;
}elsif($ARGV[0] eq "del_host"){
    &del_host;
}else{
    print "error";
    exit(0);
}

sub make_topology{
    my $width = $ARGV[1];
    my $height = $ARGV[2];
    my $sliceunit = $ARGV[3];
    my $size = $width*$height;
    my $slice_file = "test_slice.conf";
    my @mac = ("0")x12;
    my $current = 0;
    my $i;
    my $j;

    open my $fh, '>', $slice_file or die "error";
    for($i = 0; $i < $size; $i++){
	my $slicenum = $i/$sliceunit;
	if($i % $sliceunit == 0){
	    print $fh "slice".$slicenum."{";
	}
	for($j = 11; $j>=0; $j--){
	    printf $fh "%x",$mac[$j];
	    if(($j%2==0) && ($j != 0)){
		print $fh ":";
	    }
	}
	if($i%$sliceunit == $sliceunit-1 || $i == $size -1){
	    print $fh "}\n";
	}else{
	    print $fh ",";
	}
	$mac[0]++;
	while($mac[$current] == 16){
	    $mac[$current] = 0;
	    $current++;
	    $mac[$current]++;
	}
	$current = 0;
    }
    close($fh);
}

sub add_host{
    my $slice_file = "test_slice.conf";
    my $result_file = "result.conf";
    my $i;
    open my $fh, '<', $slice_file or die "error";
    open my $fh2,'>',$result_file or die "error";

    while(my $line = <$fh>){
	my $tmp = $line;
	chomp $line;
	$line =~ s/slice//g;
	my @factor = split("{",$line);
	if($factor[0] eq $ARGV[1]){
	    print $fh2 "slice".$factor[0]."{";
	    my $count = @ARGV;
	    my $mac = &convert_mac($count,$ARGV[2]);
	    print $fh2 $mac.$factor[1]."\n";
	}else{
	    print $fh2 $tmp;
	}
    }
    close($fh);
    close($fh2);
    system("mv result.conf test_slice.conf");
}

sub add_slice{
    my $slice_file = "test_slice.conf";
    my $result_file = "result.conf";
    my $i;
    open my $fh, '<', $slice_file or die "error";
    open my $fh2,'>',$result_file or die "error";

    while(my $line = <$fh>){
	print $fh2 $line;
	my $tmp = $line;
	chomp $line;
	$line =~ s/slice//g;
	my @factor = split("{",$line);
	if($factor[0] eq $ARGV[1]){
	    print "already exist\n";
	    exit(1);
	}
    }
    print $fh2 "slice".$ARGV[1]."{}\n";
    close($fh);
    close($fh2);
    system("mv result.conf test_slice.conf");
}

sub del_host{
    my $slice_file = "test_slice.conf";
    my $result_file = "result.conf";
    my $i;
    my $j;
    open my $fh, '<', $slice_file or die "error";
    open my $fh2,'>',$result_file or die "error";

    while(my $line = <$fh>){
	my $tmp = $line;
	chomp $line;
	$line =~ s/slice//g;
	my @factor = split("{",$line);
	if($factor[0] eq $ARGV[1]){
	    print $fh2 "slice".$factor[0]."{";
	    my $count = @ARGV;
	    my $mac = &convert_mac($count,$ARGV[2]);
	    my @macs = split(",",$mac);
	    my $countA = @macs;
	    my @address = split("[,}]",$factor[1]);
	    my $countB = @address;
	    for($i = 0; $i < $countB; $i++){
		for($j=0; $j < $countA; $j++){
		    if($macs[$j] eq $address[$i]){
			last;
		    }
		}
		if($j == $countA){
		    print $fh2 $address[$i];
		    if($i != $countB -1){
			print ",";
		    }
		}
	    }
	    print $fh2 "}\n";
	}else{
	    print $fh2 $tmp;
	}
    }
    close($fh);
    close($fh2);
    system("mv result.conf test_slice.conf");
}

sub del_slice{
    my $slice_file = "test_slice.conf";
    my $result_file = "result.conf";
    my $i;
    open my $fh, '<', $slice_file or die "error";
    open my $fh2,'>',$result_file or die "error";

    while(my $line = <$fh>){
	my $tmp = $line;
	chomp $line;
	$line =~ s/slice//g;
	my @factor = split("{",$line);
	if($factor[0] eq $ARGV[1]){
	}else{
	    print $fh2 $tmp;
	}
    }
    close($fh);
    close($fh2);
    system("mv result.conf test_slice.conf");
}

sub convert_mac{
    my $y = $_[0];
    my $x = 0;
    my $i = 0;
    my $j = 0;
    my $result = "";

    for($i = 2; $i < $y; $i++){
	my $x = $ARGV[$i];
	$x =~ s/host//g;
	$x--;
	my @mac = ("0")x12;
	$mac[0] = $x;
	while($mac[$j] > 16){
	    while($mac[$j] > 16){
		$mac[$j+1]++;
		$mac[$j] -= 16;
	    }
	    $j++;
	}
	
	for($j = 11; $j>=0; $j--){
	    my $tmp = sprintf("%x",$mac[$j]);
	    $result = $result.$tmp;
	    if(($j%2==0) && ($j != 0)){
		$result = $result.":";
	    }
	}
	$result = $result.",";
    }
    return $result;
}
